### DISCUSSION TITLE
want Change colour of  hind vowel diacritic  .

### DISCUSSION THREAD

**Discussion by aleem75321** [pymupdf/PyMuPDF#3175] [Looking for help]
Hi,

I want to change the colour of the hind vowel diacritic. I read the {https://github.com/pymupdf/PyMuPDF/discussions/1532) Discussion and it kind of works for me only issue is they are unable to remove old diacritic. they just overlap it new over old.

my target vowel diacritic "ं". want to change colour only. 
I tried both dict and rawdict methods both did not work.

I tried to search first and then change also did not work. when I tried to search they searched all diacritics on a page.
doc=fitz.open("test_pages/11022024_mtm_mp_03_1_col_r1.pdf")
page=doc[0]

search=page.search_for("ं")
print(search)

output:
[Rect(42.94731521606445, 127.79308319091797, 42.94731521606445, 141.1910858154297), Rect(501.72882080078125, 124.0313720703125, 501.72882080078125, 147.93136596679688), Rect(172.7761993408203, 306.5633544921875, 172.7761993408203, 371.8208312988281),....]

 but after extraction text, it failed it gave an empty 
for search in search:# loop because I have may tow many result   
    blocks=page.get_text("dict",clip=search)["blocks"]
    print(blocks)
[]
[]
[]

Method 2:- Without search, they find and change colour but do not remove old  

import fitz
import pandas as pd
from pathlib import Path
import numpy as np
doc=fitz.open("test_pages/11022024_mtm_mp_03_1_col_r1.pdf")
page=doc[0]
black = fitz.pdfcolor["black"]
blocks=page.get_text("rawdict",flags=11)["blocks"]
for block in  blocks:
    for lines in block["lines"]:
        for span in lines["spans"]:
            for char in span["chars"]:
                if char["c"] == "ं" and span["color"]==15539236:
                    x0, y0, x1, y1 = span["bbox"]  # Extract exact bounding box coordinates
                    rect = fitz.Rect(x0, y0, x1, y1)
                    print(char) 
                    # Remove original text
                    # re-insert same text - different color
                    font = fitz.Font(fontname=span['font'],fontfile=f"{span['font']}.ttf")  # this must be known somehow - or simply try some font else
                    page.add_redact_annot(rect,fontname=font,align=char["origin"])
                    tw = fitz.TextWriter(page.rect, color=black)
                    tw.append(char["origin"], text=char["c"], font=font,fontsize=span['size'])
                    tw.write_text(page)  

# Apply redactions after all text replacements

page.apply_redactions() 

# Saving Option 
out_filename=Path('out') /"11022024_mtm_mp_03_2_col_r2.pdf"

# save to a new PDF 
doc.ez_save(out_filename)

original PDF
[11022024_mtm_mp_01_1_col_r1.pdf](https://github.com/pymupdf/PyMuPDF/files/14321493/11022024_mtm_mp_01_1_col_r1.pdf)

after PDF 
[11022024_mtm_mp_03_2_col_r2.pdf](https://github.com/pymupdf/PyMuPDF/files/14321495/11022024_mtm_mp_03_2_col_r2.pdf)

**Comment by JorjMcKie** [comment]
We have noticed this already and submitted a bug report to MuPDF, please see [here](https://bugs.ghostscript.com/show_bug.cgi?id=707560).

**Reply by aleem75321** [reply]
Thanks For your quick response. My issue is different. I just want to change the colour of this  I can do that but Unable to remove the old one.  You can see in the below screenshot black colour diacritic is showing but in the background it showing old red  aslo 


![Screenshot 2024-02-18 131148](https://github.com/pymupdf/PyMuPDF/assets/69341817/5795641d-34f2-4443-85ee-3d3d39e8e99f)

**Comment by JorjMcKie** [comment]
Your problem does go back to the issue I think:
Redaction logic does not completely remove text in some complex script systems like Devanagari: whatever artifacts remain occasionally, the redaction rectangle is not completely blanked out.

**Reply by aleem75321** [reply]
Okay sir, Do you have any workaround for this? I just need to hide or remove red and imposed black

**Reply by JorjMcKie** [reply]
Additional comments:
Writing text in Devanagari using TextWriter often will not work because it is a character-based output. Devanagari however contains a complex ligatures system requiring a complex output handling called "text shaping" - as provided by HarfBuzz.
You may be fine if your text happens to contain no such complications. Otherwise I recommend using `page.insert_htmlbox()` instead, which does use HarfBuzz and provides a lot of other comfort. Among its features is automatic shrinking of content to guarantee a fit in.

**Reply by JorjMcKie** [reply]
For the time being, I suggest to fill the redaction rectangle with white (respectively the correct background color) as a workaround.

**Reply by aleem75321** [reply]
Sound good redaction rectangle with white  can you give some simple code

**Reply by JorjMcKie** [reply]
before writing the new text, do `page.draw_rect(rect, fill=fitz.pdfcolor["white"])`. Actually, doing `add_redact_annot(rect, fill=fitz.pdfcolor["white"])` should also work.

**Reply by aleem75321** [reply]
@JorjMcKie  First a big thanks for your response.

As you suggested I tried insert_htmlbox but I got challenged in bbox. The a huge difference between the Inserted text position and the real text position> and I used the same font embedded in the PDF.

As you see I need to change the colour of  " ं" in the headline and sub-headline not in body.  so I filter by font size. 
 
 

![image](https://github.com/pymupdf/PyMuPDF/assets/69341817/4f8887fa-fcf6-49db-8338-7a6dacd362d2)

doc=fitz.open("test_pages/CCI_PAGE.pdf")
page=doc[0]
arch = fitz.Archive(r"C:\Git_Repository\dot_replacement")
black = fitz.pdfcolor["red"]
blocks=page.get_text("rawdict",flags=fitz.TEXTFLAGS_TEXT)["blocks"]  
for b in  blocks:
    for l in b["lines"]:      
        for s in l["spans"]:
            if  s["size"]>12 and s['color']==2236191:
                for c in s["chars"]:
                    if c["c"] == "ं":
                        # Extract exact bounding box coordinates
                        origin = fitz.Point(c["origin"])
                        bbox = fitz.Rect(c["bbox"])
                        # print(c) 
                        page.clean_contents()
                        # page.wrap_contents()
                        # Remove original text
                        # re-insert same text - different color
                        font = fitz.Font(fontname=s['font'],fontfile=f"{s['font']}.ttf")  # this must be known somehow - or simply try some font else
                        page.add_redact_annot(c["bbox"])
                        x0,y0,x1,y1=c["bbox"]
                        x1=x0+10 # I am getting the same position of x0 and x1 which lead to an error hence a many minor changes in his coordinate                     
                        css=f"""@font-face {{font-family: {s['font']}; src: url({s['font']}.ttf);}}* {{font-family: {s['font']};font-size:5px;color:red;}}"""
                        page.insert_htmlbox((x0,y0,x1,y1),c['c'],css=css,archive=arch)
                        # tw = fitz.TextWriter(page.rect)
                        # tw.append(origin, text=c["c"], font=font,fontsize=s['size'],language="mr")
                        # tw.write_text(page,color=black)  
                   

# Apply reactions after all text replacements   
page.apply_redactions(images=fitz.PDF_REDACT_IMAGE_NONE) 

# Saving Option 
out_filename=Path('out') /"CCI_PAGE.pdf"

# save to a new PDF 
doc.ez_save(out_filename,garbage=4,deflate=True)

[CCI_page.pdf](https://github.com/pymupdf/PyMuPDF/files/14349223/CCI_page.pdf)

**Reply by JorjMcKie** [reply]
Yeah I feared that we were talking past each other.
The only thing that **_might_** help is (1) delete the **_complete_** old text via redaction (2) if artifacts remain (as discussed earlier), draw a white rectangle over the redact area, (3) **_rewrite the complete text (!)_** with desired different colors. This should be possible on a by-character level thanks to HTML.
There is no way to change the color of a single character by rewriting just it. An impossible task especially for these East-Asian fonts!

**Reply by aleem75321** [reply]
Sir, I am unable to use redaction because drawing a white rectangle over the redact area is not happening > if this happens my issue is resolved. 

can you help me to drawing  white rectangle

**Comment by JorjMcKie** [comment]
Here you are:
[test.zip](https://github.com/pymupdf/PyMuPDF/files/14349810/test.zip)

**Reply by aleem75321** [reply]
Sir very thanks for the quick reply 


I need to draw a rectangle where I mark in the Image I can draw in between characters.



![image](https://github.com/pymupdf/PyMuPDF/assets/69341817/88b4fe2b-9cbc-45b6-ab68-4a1cbf2c66d2)

**Reply by aleem75321** [reply]
I think i can do some play with ascender,descender what you suggested

**Comment by aleem75321** [comment]
@JorjMcKie Hi sir, In our previous discussion, you mention you already report a bug https://bugs.ghostscript.com/show_bug.cgi?id=707560. So I have tracked this as a bug. now this bug is resolved,  But still, I am not able to Redaction

**Reply by JorjMcKie** [reply]
what do you mean: still text stuff surviving a redaction?
If yes: an example please!

**Reply by aleem75321** [reply]
I have tried to redact it do not is not perfect for example they removed the dot and other characters also you can check the below image.

Note:- In the image yellow circle show actual place of dote and purple circle show remove other character 
![image](https://github.com/pymupdf/PyMuPDF/assets/69341817/d6f8a89d-b25f-4bdf-9c45-24a3d8b6d7e6)

**Reply by JorjMcKie** [reply]
I wish you would provide data usable for reproducing the problem. A picture alone tells me close to nothing.

I also do not understand your comment: do you mean you wanted to redact the yellow-circled stuff, but the pink one also got removed?

**Reply by aleem75321** [reply]
I'm very sorry for my previous comment. Please accept my apologies 

#Test Files I have added 2 test files 
[test.pdf](https://github.com/pymupdf/PyMuPDF/files/14865954/test.pdf)

[test1.pdf](https://github.com/pymupdf/PyMuPDF/files/14865952/test1.pdf)


Issue:-Actually, when I am deleting a word, some other words are also getting deleted along with it.
The area is in the yellow circle in that photo. That word on there is the word which I had to delete and in the purple circle is the word which has been deleted, but it should not have been deleted.




Code:-

# Iterate over each page of the document
for page in doc:
    # Find all instances of "dot" on the current page
    instances = page.search_for("ं")

    # Redact each instance of "dot" on the current page
    for inst in instances:
        text=page.get_textpage(clip=inst)
        # print(text.extractWORDS())
        page.add_redact_annot(inst,fill='red')

    # Apply the redactions to the current page
    page.apply_redactions()

# Save the modified document
doc.save('out/redacted_document.pdf')
doc.close()# Iterate over each page of the document
for page in doc:
    # Find all instances of "dot" on the current page
    instances = page.search_for("ं")

    # Redact each instance of "dot" on the current page
    for inst in instances:
        text=page.get_textpage(clip=inst)
        # print(text.extractWORDS())
        page.add_redact_annot(inst,fill='red')

    # Apply the redactions to the current page
    page.apply_redactions()

# Save the modified document
doc.save('out/redacted_document.pdf')
doc.close()

**Reply by JorjMcKie** [reply]
Thats ok, never mind.

For mostly legal reasons, redactions will remove any character that intersects the redact rectangle. This is what is happening here.
A completely different problem than before.
You have a number of choices:
1. before searching, execute `fitz.TOOLS.set_small_glyph_heights(True)`. This will reduce the height of a character bbox to the visible part (roughly).
2. If insufficient, you can fall back to some horizontal stripe part of the search hit rectangle, e.g. 20% of original height, located around the middle line.

**Reply by aleem75321** [reply]
fitz.TOOLS.set_small_glyph_heights(True) does not work for me. can you help me show some examples of point 2 How I get 20% of original height, located around the middle line. .

**Reply by aleem75321** [reply]
@JorjMcKie  As you know I try to change the colour of the dots in the headline or sub-headline. I use rawdict method which accurately information then I try to write a new dot with the desired colour as you see in the image some dots don't overlap properly so I try to redact them but as you know I face issues in this also. 

My task is only to change the colour of the dots in the PDF anyhow.

if you have any suggestions please give me some. 

#Code
import fitz
from pathlib import Path
file_path=Path(r"test_pages/test1.pdf")
doc=fitz.open(file_path)
page=doc[0]
#Set Colour for outoput PDF
black = fitz.pdfcolor["red"]
page.clean_contents()
# Extracting text Form PDF 
fitz.TOOLS.set_small_glyph_heights(True)
blocks=page.get_text("rawdict",flags=fitz.TEXTFLAGS_TEXT,sort=True)["blocks"]  
count=0
for b in  blocks:
    for l in b["lines"]:  
        for s in l["spans"]:
            if s["size"]>15 and s['color']==2236191:
                for c in s["chars"]:
                    if c["c"] == "ं":
                        count=count+1
                      
                        # Extract exact bounding box coordinates
                        a = s["ascender"]
                        d = s["descender"]
                
                        x1,y1,x2,y2=c["bbox"]
                       
                        # page.add_redact_annot(c['bbox'],fill='red')

                        o = fitz.Point(c["origin"])  # its y-value is the baseline
                        # #Calculating font 
                        font_size=((o.y - s["size"] * d / (a - d))+a)-(o.y - s["size"] * d / (a - d) - s["size"]+d)
                        font_size=(round(font_size,0))                       
                        
                        ## re-insert same text - different color
                        font = fitz.Font(fontname=s['font'],fontfile=f"{s['font']}.ttf")  # this must be known somehow - or simply try some font else

                        tw = fitz.TextWriter(page.rect)
                        tw.append(c["origin"], text=c["c"], font=font,fontsize=font_size,language="mr")
                        tw.write_text(page,color=black)  
                                
# Apply reactions after all text replacements   
page.apply_redactions(images=fitz.PDF_REDACT_IMAGE_NONE) 

#Saving Backup File furture use 
out_fpath="OUT/"+file_path.stem+".pdf"
doc.save(out_fpath,garbage=3, deflate=True)

#Images
![image](https://github.com/pymupdf/PyMuPDF/assets/69341817/c6128262-c569-49be-93cd-e8aa249a3c96)
#PDF
[test1.pdf](https://github.com/pymupdf/PyMuPDF/files/14898120/test1.pdf)
